### PR TITLE
Add electron autoupdater

### PR DESCRIPTION
**Pull request by thundron**
Change-type: patch
Changelog-entry: Add support for auto-updating feature
Signed-off-by: Giovanni Garufi <giovanni@balena.io>

### CODE REVIEW

File: lib/gui/etcher.js [hunk 1]
```diff
@@ -18,14 +18,36 @@
...
+/**
+ *
+ * @param {Object} autoUpdaterConfig - electron-updater configs
+ * @example checkForUpdates()
+ */
+const checkForUpdates = async (autoUpdaterConfig) => {
+  try {
+    _.merge(autoUpdater, autoUpdaterConfig)
+    const release = await autoUpdater.checkForUpdates()
+    if (release.updateInfo.stagingPercentage) {
```
[the remark this answers is not in the document]
Comment (thundron) [reply lib/gui/etcher.js]: It's in order to trigger updates only on releases that have a stagingPercentage prop defined as @nazrhom mentioned on fd

File: lib/gui/etcher.js [hunk 2]
```diff
@@ -83,6 +105,28 @@ const createMainWindow = () => {
...
+    })
+    if (settings.get('updatesEnabled')) {
+      try {
+        const onlineConfig = await getConfig(configUrl)
+        const autoUpdaterConfig = _.get(onlineConfig, [ 'autoUpdates', 'autoUpdaterConfig' ], {
+          autoDownload: false
+        })
+        await checkForUpdates(autoUpdaterConfig)
+        // eslint-disable-next-line no-magic-numbers
+        const CHECK_FOR_UPDATES_TIMER = _.get(onlineConfig, [ 'autoUpdates', 'checkForUpdatesTimer' ], 300000)
```
[the remark this answers is not in the document]
Comment (thundron) [reply lib/gui/etcher.js]: It was originally an env var, so 🤷‍♂️

File: lib/gui/etcher.js [hunk 3]
```diff
@@ -18,14 +18,39 @@
...
 const config = settings.getDefaults()
 
+const configUrl = settings.get('configUrl') || 'http://balena.io/etcher/static/config.json'
+
+/**
+ *
+ * @param {Number} timer - timer to wait to check for updates
+ * @example checkForUpdates()
+ */
+const checkForUpdates = async (timer) => {
```
Comment (zvin) [inline lib/gui/etcher.js]: `interval` would be more appropriate than `timer`

File: lib/gui/etcher.js [hunk 4]
```diff
@@ -83,6 +108,28 @@ const createMainWindow = () => {
...
+    if (settings.get('updatesEnabled')) {
+      try {
+        const onlineConfig = await getConfig(configUrl)
+        const autoUpdaterConfig = _.get(onlineConfig, [ 'autoUpdates', 'autoUpdaterConfig' ], {
+          autoDownload: false
+        })
+        _.merge(autoUpdater, autoUpdaterConfig)
+        // eslint-disable-next-line no-magic-numbers
+        const checkForUpdatesTimer = _.get(onlineConfig, [ 'autoUpdates', 'checkForUpdatesTimer' ], 300000)
+        await checkForUpdates(checkForUpdatesTimer)
```
Comment (zvin) [inline lib/gui/etcher.js]: No need to await here, this is an infinite loop.

File: lib/gui/etcher.js [hunk 5]
```diff
@@ -18,14 +18,39 @@
...
+ *
+ * @param {Number} timer - timer to wait to check for updates
+ * @example checkForUpdates()
+ */
+const checkForUpdates = async (timer) => {
+  while (true) {
+    try {
+      const release = await autoUpdater.checkForUpdates()
+      if (release.updateInfo.stagingPercentage) {
+        await autoUpdater.downloadUpdate(release.cancellationToken)
```
Comment (zvin) [inline lib/gui/etcher.js]: Why is `release.cancellationToken` needed?

Comment (thundron) [reply lib/gui/etcher.js]: It's used to cancel the download promise by `electron-updater`

**Comment by thundron** [comment]
@resin-ci retest

**Comment by thundron** [comment]
@resin-ci retest

**Comment by thundron** [comment]
@resin-ci retest

**Pull request merged**